src/window.rs

```diff
@@ -151,37 +196,25 @@ fn retransmit_all_leader_blocks(
     consumed: u64,
     received: u64,
     retransmit: &BlobSender,
+    window: &SharedWindow,
 ) -> Result<()> {
-    let mut retransmit_queue = VecDeque::new();
+    let mut retransmit_queue: VecDeque<SharedBlob> = VecDeque::new();
     if let Some(leader) = maybe_leader {
+        let leader_id = leader.id;
         for b in dq {
-            let p = b.read().expect("'b' read lock in fn recv_window");
-            //TODO this check isn't safe against adverserial packets
-            //we need to maintain a sequence window
-            let leader_id = leader.id;
-            trace!(
-                "idx: {} addr: {:?} id: {:?} leader: {:?}",
-                p.get_index().expect("get_index in fn recv_window"),
-                p.get_id().expect("get_id in trace! fn recv_window"),
-                p.meta.addr(),
-                leader_id
-            );
-            if p.get_id().expect("get_id in fn recv_window") == leader_id {
-                //TODO
-                //need to copy the retransmitted blob
-                //otherwise we get into races with which thread
-                //should do the recycling
-                //
-                //a better abstraction would be to recycle when the blob
-                //is dropped via a weakref to the recycler
-                let nv = recycler.allocate();
-                {
-                    let mut mnv = nv.write().expect("recycler write lock in fn recv_window");
-                    let sz = p.meta.size;
-                    mnv.meta.size = sz;
-                    mnv.data[..sz].copy_from_slice(&p.data[..sz]);
+            add_block_to_retransmit_queue(b, leader_id, recycler, &mut retransmit_queue);
+        }
+
+        for w in window
```

Comment:
we look at all the blobs in the window every time we call retransmit_all_leader_blocks()?

Reply:
Yes, we need to match the senders ID with the leader's and retransmit only if it matches.
Maybe it'll help if SharedWindow had a state indicating whether the window has any blobs that have leader_unkown = true. The code can loop thru the window only when that state is true.

Reply:
or you could keep a bit of local state, similar to consumed, that says whether there are blobs that might need retransmission

Reply:
That's much better, and localized. Thanks @rob-solana

Reply:
leader_unknown can move from false->true->false->true?  probably want tests that cover this.  do I read right that this change increases node count by 80%?

Reply:
I am not certain about exact % increase in node count. Earlier it was more conservatively set to 100. I think 120 nodes were also going thru (sometimes higher also), but it was flaky. I am able to run 180 nodes pretty consistently now.

I am looking into adding the test.